src/main/java/gyro/aws/codebuild/CodebuildProjectEnvironment.java

```diff
@@ -0,0 +1,226 @@
+        "ARM_CONTAINER",
+        "WINDOWS_SERVER_2019_CONTAINER" })
+    public String getType() {
+        return type;
+    }
+
+    public void setType(String type) {
+        this.type = type;
+    }
+
+    /**
+     * The certificate used with the build project.
+     */
+    @Updatable
+    public String getCertificate() {
+        return certificate;
+    }
+
+    public void setCertificate(String certificate) {
+        this.certificate = certificate;
+    }
+
+    /**
+     * The list of environment variables available to builds for the build project.
+     */
+    @Updatable
+    public List<CodebuildProjectEnvironmentVariable> getEnvironmentVariables() {
+        return environmentVariables;
+    }
+
+    public void setEnvironmentVariables(List<CodebuildProjectEnvironmentVariable> environmentVariables) {
+        this.environmentVariables = environmentVariables;
+    }
+
+    /**
+     * The type of credentials the build project uses to pull images in the build.
+     */
+    @Updatable
+    @ValidStrings({ "CODEBUILD", "SERVICE_ROLE" })
+    public String getImagePullCredentialsType() {
+        return imagePullCredentialsType;
+    }
+
+    public void setImagePullCredentialsType(String imagePullCredentialsType) {
+        this.imagePullCredentialsType = imagePullCredentialsType;
+    }
+
+    /**
+     * The field that specifies running the Docker daemon inside a Docker container.
+     */
+    @Updatable
+    public Boolean getPrivalegedMode() {
+        return privalegedMode;
+    }
+
+    public void setPrivalegedMode(Boolean privalegedMode) {
+        this.privalegedMode = privalegedMode;
+    }
+
+    /**
+     * The credentials for access to a private registry.
+     */
+    @Updatable
+    public CodebuildRegistryCredential getRegistryCredential() {
+        return registryCredential;
+    }
+
+    public void setRegistryCredential(CodebuildRegistryCredential registryCredential) {
+        this.registryCredential = registryCredential;
+    }
+
+    @Override
+    public void copyFrom(ProjectEnvironment model) {
+        setComputeType(model.computeTypeAsString());
+        setImage(model.image());
+        setType(model.typeAsString());
+        setCertificate(model.certificate());
+        setImagePullCredentialsType(model.imagePullCredentialsTypeAsString());
+        setPrivalegedMode(model.privilegedMode());
+
+        if (model.environmentVariables() != null) {
+            List<CodebuildProjectEnvironmentVariable> environmentVariables = new ArrayList<>();
+
+            CodebuildProjectEnvironmentVariable environmentVariable = newSubresource(
+                CodebuildProjectEnvironmentVariable.class);
+
+            for (EnvironmentVariable variable : model.environmentVariables()) {
+                environmentVariable.copyFrom(variable);
+                environmentVariables.add(environmentVariable);
+            }
+
+            setEnvironmentVariables(environmentVariables);
+        }
```

Comment:
1. Whenever you are setting something conditionally in the copyFrom, always clear / set it to null, either to begin with, or as an else clause. The refresh when its being called gets all the field values from the state file of the previous run, so post refresh it is your responsibility to not only set new values but unset the old ones if necessary.

2. Using stream/lambda coding syntax to reduce code lines and variables while keeping the readability. 

```suggestion
        getEnvironmentVariables().clear();
        if (model.environmentVariables() != null) {
                model.environmentVariables().forEach(variable -> {
    	                CodebuildProjectEnvironmentVariable environmentVariable = newSubresource(
        	                CodebuildProjectEnvironmentVariable.class);
                        environmentVariable.copyFrom(variable);
                        getEnvironmentVariables().add(environmentVariable)
                });
        }
```